File: src/main/java/io/alexjoest/stackupup/core/CompatibilityLimitPatch.java
package io.alexjoest.stackupup.core;

import io.alexjoest.stackupup.Constants;
import io.alexjoest.stackupup.StackUpUpIds;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.tree.AbstractInsnNode;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.InsnList;
import org.objectweb.asm.tree.IntInsnNode;
import org.objectweb.asm.tree.MethodInsnNode;
import org.objectweb.asm.tree.MethodNode;

import java.util.Collections;
import java.util.List;
import java.util.ListIterator;
import java.util.function.Consumer;

public final class CompatibilityLimitPatch {

    private CompatibilityLimitPatch() {}

    private static final int DEFAULT_STACK_LIMIT = Constants.VANILLA_STACK_LIMIT;
    private static final String HELPER_OWNER = StackUpUpIds.STACK_LIMIT_HOOKS_INTERNAL_NAME;
    private static final String HELPER_NAME = "getCompatibilityStackSize";
    private static final String HELPER_DESC = "()I";

    public static List<Consumer<ClassNode>> planFor(String transformedName) {
        return planFor(transformedName, null);
    }

    public static List<Consumer<ClassNode>> planFor(String transformedName, byte[] basicClass) {
        int declaredProfiles;
        if (basicClass != null) {
            declaredProfiles = DynamicCompatMethodProbe.detectProfiles(basicClass);
        } else {
            declaredProfiles = DynamicCompatTargetProfile.NONE;
        }

        if (declaredProfiles == DynamicCompatTargetProfile.NONE) {
            return Collections.emptyList();
        }

        int profile = DynamicCompatTargetClassifier.classify(transformedName, declaredProfiles);
        if (profile == DynamicCompatTargetProfile.ITEM_HANDLER) {
            return Collections.emptyList();
        }

        String[] methods = DynamicCompatTargetProfile.methodsFor(profile);
        if (methods == null) {
            return Collections.emptyList();
        }

        return Collections.singletonList(rewrite(methods));
    }

    static Consumer<ClassNode> rewrite(String... methods) {
        return node -> {
            for (MethodNode method : node.methods) {
                if (!matchesAny(method.name, methods)) {
                    continue;
                }

                InsnList instructions = method.instructions;
                ListIterator<AbstractInsnNode> iterator = instructions.iterator();
                while (iterator.hasNext()) {
                    AbstractInsnNode instruction = iterator.next();
                    if (instruction.getOpcode() != Opcodes.BIPUSH) {
                        continue;
                    }

                    IntInsnNode intInstruction = (IntInsnNode) instruction;
                    if (intInstruction.operand != DEFAULT_STACK_LIMIT) {
                        continue;
                    }

                    iterator.set(new MethodInsnNode(
                        Opcodes.INVOKESTATIC,
                        HELPER_OWNER,
                        HELPER_NAME,
                        HELPER_DESC,
                        false
                    ));
                }
            }
        };
    }

    private static boolean matchesAny(String name, String[] candidates) {
        for (String candidate : candidates) {
            if (candidate.equals(name)) {
                return true;
            }
        }
        return false;
    }
}

File: src/main/kotlin/io/alexjoest/stackupup/core/CompatibilityLimitPatch.kt
package io.alexjoest.stackupup.core

import io.alexjoest.stackupup.Constants
import io.alexjoest.stackupup.StackUpUpIds
import org.objectweb.asm.Opcodes
import org.objectweb.asm.tree.ClassNode
import org.objectweb.asm.tree.IntInsnNode
import org.objectweb.asm.tree.MethodInsnNode
import java.util.Collections
import java.util.function.Consumer

internal object CompatibilityLimitPatch {
    private const val DEFAULT_STACK_LIMIT: Int = Constants.VANILLA_STACK_LIMIT
    private const val HELPER_OWNER: String = StackUpUpIds.STACK_LIMIT_HOOKS_INTERNAL_NAME
    private const val HELPER_NAME: String = "getCompatibilityStackSize"
    private const val HELPER_DESC: String = "()I"

    fun planFor(transformedName: String, basicClass: ByteArray? = null): List<Consumer<ClassNode>> {
        val declaredProfiles = basicClass?.let(DynamicCompatMethodProbe::detectProfiles)
        if (declaredProfiles == DynamicCompatTargetProfile.NONE) {
            return Collections.emptyList()
        }

        val profile = if (declaredProfiles != null) {
            DynamicCompatTargetClassifier.classify(transformedName, declaredProfiles)
        } else {
            DynamicCompatTargetClassifier.classify(transformedName)
        }
        val methods = DynamicCompatTargetProfile.methodsFor(profile) ?: return Collections.emptyList()

        return Collections.singletonList(rewrite(*methods))
    }

    fun rewrite(vararg methods: String): Consumer<ClassNode> {
        return Consumer { node ->
            for (method in node.methods) {
                if (!matchesAny(method.name, methods)) {
                    continue
                }

                var patchesMade = 0
                val iterator = method.instructions.iterator()
                while (iterator.hasNext()) {
                    val instruction = iterator.next()
                    if (instruction.opcode != Opcodes.BIPUSH) {
                        continue
                    }

                    val intInstruction = instruction as IntInsnNode
                    if (intInstruction.operand != DEFAULT_STACK_LIMIT) {
                        continue
                    }

                    iterator.set(
                        MethodInsnNode(
                            Opcodes.INVOKESTATIC,
                            HELPER_OWNER,
                            HELPER_NAME,
                            HELPER_DESC,
                            false
                        )
                    )
                    patchesMade++
                }
            }
        }
    }

    private fun matchesAny(name: String, candidates: Array<out String>): Boolean {
        for (candidate in candidates) {
            if (candidate == name) {
                return true
            }
        }
        return false
    }
}
